File: php/features.php
<?php
declare(strict_types=1);

// Tillo SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class TilloFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new TilloBaseFeature();
            case "debug":
                return new TilloDebugFeature();
            case "idempotency":
                return new TilloIdempotencyFeature();
            case "metrics":
                return new TilloMetricsFeature();
            case "paging":
                return new TilloPagingFeature();
            case "ratelimit":
                return new TilloRatelimitFeature();
            case "retry":
                return new TilloRetryFeature();
            case "test":
                return new TilloTestFeature();
            case "timeout":
                return new TilloTimeoutFeature();
            default:
                return new TilloBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/tillo_sdk/features.py
# Tillo SDK feature factory

from tillo_sdk.feature.base_feature import TilloBaseFeature
from tillo_sdk.feature.debug_feature import TilloDebugFeature
from tillo_sdk.feature.idempotency_feature import TilloIdempotencyFeature
from tillo_sdk.feature.metrics_feature import TilloMetricsFeature
from tillo_sdk.feature.paging_feature import TilloPagingFeature
from tillo_sdk.feature.ratelimit_feature import TilloRatelimitFeature
from tillo_sdk.feature.retry_feature import TilloRetryFeature
from tillo_sdk.feature.test_feature import TilloTestFeature
from tillo_sdk.feature.timeout_feature import TilloTimeoutFeature


_FEATURES = {
    "base": lambda: TilloBaseFeature(),
    "debug": lambda: TilloDebugFeature(),
    "idempotency": lambda: TilloIdempotencyFeature(),
    "metrics": lambda: TilloMetricsFeature(),
    "paging": lambda: TilloPagingFeature(),
    "ratelimit": lambda: TilloRatelimitFeature(),
    "retry": lambda: TilloRetryFeature(),
    "test": lambda: TilloTestFeature(),
    "timeout": lambda: TilloTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
